File: javascript/perfect-numbers/perfect-numbers.js
export const classify = (n) => {
  if (n < 1) throw new Error('Classification is only possible for natural numbers.');
  const factors = [1];
  for (let i = 2; i <= Math.sqrt(n); i++) {
    if (n % i === 0) {
      factors.push(i);
      if (i !== Math.sqrt(n)) {
        factors.push(n / i);
      }
    }
  }
  const sum = factors.reduce((acc, f) => acc + f, 0);
  if (sum < n || sum === 1) return 'deficient';
  return sum === n ? 'perfect' : 'abundant';
}

File: python/perfect-numbers/perfect_numbers.py
from math import sqrt

def classify(number):
  if number <= 0:
    raise ValueError('Number must be a natural number')
  if number == 1:
    return 'deficient'

  factors = [1]
  root = sqrt(number)
  factor = 2

  while factor <= root:
    if number % factor == 0:
      factors.append(factor)
      if (factor != root):
        factors.append(number / factor)
    factor +=1

  factor_sum = sum(factors)
  if factor_sum == number:
    return 'perfect'
  if factor_sum > number:
    return 'abundant'
  return 'deficient'